docs:#799tiered storage support for azure (#1057)
redpanda-data/documentation: docs/manage/tiered-storage.mdx

This feature requires an [Enterprise license](../../get-started/licenses). To upgrade, contact [Redpanda sales](https://redpanda.com/try-redpanda?section=enterprise-cloud).
:::

Tiered Storage helps to lower storage costs by offloading log segments to cloud storage. You can specify the amount of storage you want to retain in local storage. You don't need to specify which log segments you want to move, because Redpanda moves them automatically based on cluster-level configuration properties. Tiered Storage indexes where data is offloaded, so it can retrieve the data when you need it.

Redpanda natively supports Tiered Storage with Amazon S3, Google Cloud Storage (GCS), and Microsoft Azure Blob Storage (ABS). 

:::caution
Do not use Tiered Storage with [data transforms](../../labs/data-transform), which is a lab feature in technical preview. Lab features are not supported for production usage.
:::

The following image illustrates the Tiered Storage architecture: remote write uploads data from Redpanda to cloud storage, and remote read fetches data from cloud storage to Redpanda.

![Tiered Storage architecture](../../static/img/tiered_storage.png) 

## Set up

To use Tiered Storage:

> 1. [Configure cloud storage](../../data-management/tiered-storage/#configure-cloud-storage).

Reviewer: This link did not work for me in the Preview.

Reviewer: Actually, none of the links in this list work in Preview.

Author: thanks, @Feediver1 -- I fixed them!

2. [Enable Tiered Storage for a cluster](../../data-management/tiered-storage/#enable-tiered-storage-for-a-cluster), or [enable Tiered Storage for a topic](../../data-management/tiered-storage/#enable-tiered-storage-for-a-topic).
3. [Set retention limits](../../data-management/tiered-storage/#set-retention-limits).

### Configure cloud storage

<Tabs>
    <TabItem value="S3" label="Amazon S3" default>

> Configure access to Amazon S3 with either an IAM role attached to the instance or with access keys.

Reviewer: I notice that you are still using tabs here. You cannot link to any of the topics in them. Have you considered making the set up for each platform a subhead instead?

To configure access to an S3 bucket with an IAM role: 

> 1. Configure an [IAM role](../../manage/security/iam-roles). 

Reviewer: This link takes you to the h1 for the topic. Why not link to the heading that actually has the configuration info? https://deploy-preview-1057--redpanda-documentation.netlify.app/docs/manage/security/iam-roles/#configuring-iam-roles

Author: thanks, fixed!

2. Run the `rpk cluster config edit` command, then edit the following required properties:

   ```properties
   cloud_storage_enabled: true
   cloud_storage_credentials_source: <aws_instance_metadata>
   cloud_storage_region: <region>
   cloud_storage_bucket: <redpanda_bucket_name>
   ```
   Replace `<variables>` with your own values.

To configure access to an S3 bucket with access keys instead of an IAM role:

> 1. Grant a user the following permissions to read and create objects on the bucket to be used with the cluster (or on all buckets): `GetObject`, `DeleteObject`, `PutObject`.

Reviewer: Not sure if you were aware, but @daisukebe has a PR waiting to be merged and it added a couple other min permissions: `PutObjectTagging`, and `ListBucket`
Here is a link to his PR in case you want to take a look and see if you should add those permissions here too:
https://github.com/redpanda-data/documentation/pull/989

5. Run the `rpk cluster config edit` command, then edit the following required properties:

   ```properties
   cloud_storage_enabled: true
   cloud_storage_access_key: ***
   cloud_storage_api_endpoint: storage.googleapis.com
   cloud_storage_bucket: <redpanda_bucket_name>
   cloud_storage_region: <region>
   cloud_storage_secret_key: ***
   ```
   Replace `<variables>` with your own values.

For additional properties, see [Tiered Storage configuration properties](../tiered-storage#tiered-storage-configuration-properties).

  </TabItem>
    <TabItem value="Azure" label="Azure Blob Storage" default>

To configure access to Azure Blob Storage with shared keys:

> 1. Copy an account access key for the Azure container you want Redpanda to use. You'll enter this for the `cloud_storage_azure_shared_key` property. For information on how to view your account access keys, see the [Azure documentation](https://learn.microsoft.com/en-us/azure/storage/common/storage-account-keys-manage?toc=%2Fazure%2Fstorage%2Fblobs%2Ftoc.json&bc=%2Fazure%2Fstorage%2Fblobs%2Fbreadcrumb%2Ftoc.json&tabs=azure-portal#view-account-access-keys). 

Reviewer: ```suggestion
1. Copy an account access key for the Azure container you want Redpanda to use and enter it in the  `cloud_storage_azure_shared_key` property. For information on how to view your account access keys, see the [Azure documentation](https://learn.microsoft.com/en-us/azure/storage/common/storage-account-keys-manage?toc=%2Fazure%2Fstorage%2Fblobs%2Ftoc.json&bc=%2Fazure%2Fstorage%2Fblobs%2Fbreadcrumb%2Ftoc.json&tabs=azure-portal#view-account-access-keys). 
```

   cloud_storage_region: <region>
   cloud_storage_secret_key: ***
   ```
   Replace `<variables>` with your own values.

For additional properties, see [Tiered Storage configuration properties](../tiered-storage#tiered-storage-configuration-properties).

  </TabItem>
    <TabItem value="Azure" label="Azure Blob Storage" default>

To configure access to Azure Blob Storage with shared keys:

1. Copy an account access key for the Azure container you want Redpanda to use. You'll enter this for the `cloud_storage_azure_shared_key` property. For information on how to view your account access keys, see the [Azure documentation](https://learn.microsoft.com/en-us/azure/storage/common/storage-account-keys-manage?toc=%2Fazure%2Fstorage%2Fblobs%2Ftoc.json&bc=%2Fazure%2Fstorage%2Fblobs%2Fbreadcrumb%2Ftoc.json&tabs=azure-portal#view-account-access-keys). 
2. Run the `rpk cluster config edit` command, then edit the following required properties:

   ```properties
   cloud_storage_enabled: true
   cloud_storage_azure_shared_key: <azure_account_access_key>
   cloud_storage_azure_storage_account: <azure_account_name>
   cloud_storage_azure_container: <redpanda_container_name>

>    cloud_storage_region: <region>

Reviewer 2: The region should not be specified when using ABS.

Redpanda supports retention limits and compaction for topics using Tiered Storage. Set retention limits to purge topic data after it reaches a specified age or size.

> Starting in Redpanda version 22.3, cloud storage is the default storage tier for all streaming data, and retention properties work the same for Tiered Storage topics and local storage topics. Data is retained in the cloud until it reaches the configured time or size limit. Retention limits apply to data stored in cloud storage, even though additional data may be cached or managed in local storage.

Reviewer 3: Is the last sentence redundant with the first sentence, that retention applies for both cloud and local storage? If yes, consider removing or rewriting the last sentence. 

```suggestion
Starting in Redpanda version 22.3, cloud storage is the default storage tier for all streaming data, and retention properties work the same for Tiered Storage topics and local storage topics. Data is retained in the cloud until it reaches the configured time or size limit. Retention limits apply to both data stored in cloud storage and data cached or managed in local storage.
```

Author: Thanks @shorgi! I'll remove the last sentence.

- If `retention.ms` or `retention.bytes` is set, it initializes the `remote_retention_` properties in `topic_properties`.
- If `retention.local.target.ms` or `retention.local.target.bytes` is set, it is normalized and used to initialize the retention fields in `topic_properties`.
- If properties are not specified:

>   - Starting in release 22.3, new topics use the default retention values of 1 (local) and 7 days (cloud).

Reviewer: ```suggestion
  - Starting in version 22.3, new topics use the default retention values of 1 (local) and 7 days (cloud).
```

You can configure Redpanda to start a remote write periodically. This is useful if the ingestion rate is low and the segments are kept open for long periods of time. You specify a number of seconds for the timeout, and if that time has passed since the previous write and the partition has new data, Redpanda starts a new write. This limits data loss in the event of catastrophic failure and adds a guarantee that you only lose the specified number of seconds of data.

> Setting idle timeout to a very short interval can create a lot of small files, which can affect throughput. If you decide to set a value for idle timeout, start out with 600 seconds, which prevents the creation of so many small files that throughput is affected when you recover the files.

Reviewer 3: ```suggestion
Setting idle timeout to a very short interval can create a lot of small files, which can affect throughput. If you decide to set a value for idle timeout, start with 600 seconds, which prevents the creation of so many small files that throughput is affected when you recover the files.
```

Reviewer: +1

Use the `cloud_storage_segment_max_upload_interval_sec` property to set the number of seconds for idle timeout. If this property is empty, Redpanda uploads metadata to cloud storage, but the segment is not uploaded until it reaches the `segment.bytes` size. Edit the property value using the `rpk cluster config edit` command. By default, the property is empty.

### Reconciliation 

> Reconciliation is a Redpanda process that runs periodically on every node. It monitors partitions and decides which partitions are uploaded on each node to guarantee that data is uploaded only once. It also balances the workload evenly between nodes. 

Reviewer 3: ```suggestion
Reconciliation is a Redpanda process that monitors partitions and decides which partitions are uploaded on each node to guarantee that data is uploaded only once. It runs periodically on every node. It also balances the workload evenly between nodes. 
```

Setting idle timeout to a very short interval can create a lot of small files, which can affect throughput. If you decide to set a value for idle timeout, start out with 600 seconds, which prevents the creation of so many small files that throughput is affected when you recover the files.

Use the `cloud_storage_segment_max_upload_interval_sec` property to set the number of seconds for idle timeout. If this property is empty, Redpanda uploads metadata to cloud storage, but the segment is not uploaded until it reaches the `segment.bytes` size. Edit the property value using the `rpk cluster config edit` command. By default, the property is empty.

### Reconciliation 

Reconciliation is a Redpanda process that runs periodically on every node. It monitors partitions and decides which partitions are uploaded on each node to guarantee that data is uploaded only once. It also balances the workload evenly between nodes. 

> Redpanda reacts to partition leadership changes. When leadership rebalancing occurs, the reconciliation process could start uploading to cloud storage from a different node.

Reviewer 2: This makes it sound a bit like reconciliation is an extra process that happens within the tiered storage mechanism. We simply really on the leadership notifications to change the node that's uploading, so there's no reconciliation really.

The upload controller prevents Redpanda from running out of disk space. If `remote.write` is set to `true`, Redpanda cannot evict log segments that have not been uploaded to cloud storage. If the remote write process cannot keep up with the amount of data that needs to be uploaded to cloud storage, the upload controller increases priority for the upload process. The upload controller measures the size of the upload periodically and tunes the priority of the remote write process. 

## Remote read

Remote read fetches data from cloud storage using the Kafka API. Use remote read in conjunction with remote write to enable Tiered Storage. If you use remote read without remote write, there is nothing for Redpanda to read.

Normally, when data is evicted locally, it is no longer available. If the consumer starts consuming the partition from the beginning, the first offset is the smallest offset available locally. However, if Tiered Storage is enabled with the `redpanda.remote.read` and `redpanda.remote.write` flags, the data is always uploaded to remote storage before it's deleted. This guarantees that the data is always available either locally or remotely.

> When data is available remotely and Tiered Storage is enabled, the client can start consuming data from offset 0, even if the data is no longer stored locally. 

Reviewer 2: Starting from offset 0 is not a requirement. Clients can start from anywhere.

Reconciliation is a Redpanda process that runs periodically on every node. It monitors partitions and decides which partitions are uploaded on each node to guarantee that data is uploaded only once. It also balances the workload evenly between nodes. 

:::note

> After partition leadership rebalancing, Redpanda could change the node that's uploading to cloud storage.

Reviewer 2: Maybe I'm being pedantic, but I don't think this reflects the fact that the node uploading to cloud storage is *always* the leader. Hence, whenever the leader for a partition changes, uploads will proceed from that node.

Author: Thanks @VladLazar! Does this sounds correct:

Reviewer 2: That sounds good to me. I'd go for `will stop uploads` instead of `could stop uploads`, but I don't mind too much.

rpk topic create <topic_name> -c redpanda.remote.recovery=true
```

To create a new topic using remote recovery and enable Tiered Storage on the new topic with the `redpanda.remote.write` and `redpanda.remote.read` flags:

```bash
rpk topic create <topic_name> -c redpanda.remote.recovery=true -c redpanda.remote.write=true -c redpanda.remote.read=true
```

## Retries and backoff 

> If the cloud provider replies with an error message that the server is busy, Redpanda retries the request. Redpanda always uses exponential backoff with cloud connections. Redpanda retries the request if it receives any of the following errors:

Reviewer 2: I'm not too sure how accurate this list is anymore and it also differs slightly between Azure and AWS. Should we maybe remove this list and say that Redpanda figures out which errors to retry and and which not to?

| Property      | Description |
| ----------- | ----------- |
| `cloud_storage_enabled` | Global flag that enables Tiered Storage. <br/>Set to true to enable Tiered Storage. Default is false.  |
| `cloud_storage_region` | Cloud storage region. <br/>Required for AWS and GCS. |
| `cloud_storage_bucket` | AWS or GCS bucket name. <br/>Required for AWS and GCS. |
| `cloud_storage_credentials_source` | AWS or GCS instance metadata. <br/>Required for AWS and GCS with IAM roles. |
| `cloud_storage_access_key` | AWS or GCS access key. <br/>Required for AWS and GCS with access keys. |
| `cloud_storage_secret_key` | AWS or GCS secret key. <br/>Required for AWS and GCS with access keys.  |
| `cloud_storage_api_endpoint` | AWS or GCS API endpoint. <br/>- For AWS, this can be left blank. It’s generated automatically using the region and bucket. <br/>- For Google Cloud Service, use `storage.googleapis.com`. |
| `cloud_storage_azure_container` | Azure container name. <br/>Required for ABS. |
| `cloud_storage_azure_storage_account` | Azure account name. <br/>Required for ABS. |
| `cloud_storage_azure_shared_key` | Azure shared key. <br/>Required for ABS. |

> | `cloud_storage_cache_size` | Maximum size of the disk cache used by Tiered Storage. Default is 20GiB. |

Reviewer: ```suggestion
| `cloud_storage_cache_size` | Maximum size of the disk cache used by Tiered Storage. Default is 20 GiB. |
```

> In addition, you might want to change the following property in the `redpanda.yaml` file of each node:

Reviewer: You might want to change...why might you? This suggests users can change it, but is vague about when and why you would do so. Would be helpful to throw in one or two reasons someone would/should change it.

| Property      | Description |
| ----------- | ----------- |

> | `cloud_storage_max_connections` | The maximum number of connections to cloud storage on a node for each CPU. Remote read and remote write share the same pool of connections. This means that if a connection is used to upload a segment, it cannot be used to download another segment. If this value is too small, some workloads might starve for connections, which results in delayed uploads and downloads. If this value is too large, Redpanda tries to upload a lot of files at the same time and might overwhelm the system. Default is 20. |

Reviewer: It would be helpful to readers to get a sense of what you mean by "too small" and "too large" here. maybe:
If this value is too small (for example x or fewer), some workloads might starve for connections, which results in delayed uploads and downloads. If this value is too large (x or more), Redpanda tries to upload a lot of files at the same time and might overwhelm the system.

Fill in the x value of what is small and large.
